### ISSUE TITLE
Add the ability to "get" the OSPlatform

### ISSUE THREAD

**Issue by isaacabraham** [dotnet/runtime#1883]
You can check the OSPlatform via the `System.Runtime.InteropServices.IsOSPlatform` method, but it would be very useful to have an actual property on the `System.Runtime.InteropServices` in addition to a method.

This would lend itself well for features such as pattern matching rather than if / then expressions or statements, which at least have something more in the way of exhaustiveness checks e.g.

```fsharp
open System.Runtime

match InteropServices.OSPlatform with
| OSPlatform.Windows -> ...
| OSPlatform.Linux -> ...
| _ -> ...
```

In F# you can sort of work around this with active patterns but it's not ideal, and in C# I suppose it'd be more work.

**Comment by vcsjones** [comment]
The `OSPlatform` type is not an enum - it's a class with static properties. Even if you do have an instance of an `OSPlatform` it's still difficult to pattern match on. e.g. this doesn't work.

```fsharp
let plat = OSPlatform.Create "DOS"
let name = match plat with
           | OSPlatform.Linux -> "Linux"
           | OSPlatform.Windows -> "Windows"
           | _ -> "Other"

printfn "Platform: %s" name
```

So I am curious if this is workable, or perhaps I'm too rusty with F# to see how this could pattern match well.

**Comment by isaacabraham** [comment]
Yep, I know it's not an enum. And you're right, it can never be exhaustive without a wildcard handler (even an enum can't be exhaustive as you can happily cast an int to it). However, there appear to be three "standard" values in the library (Windows, Linux and OSX) which I imagine the majority of people will test against.

These sorts of tests normally would be done through pattern matching rather than if / thens - it's just the idiomatic way of doing things - but without a value to match on, it's a bit awkward.

I also found it a little unusual (IMHO, but maybe this is more design aesthetics) that I couldn't ask the runtime "What OS are you?". Instead you have to ask "Are you this OS? Are you that OS? Or the other OS"?.

I'd be happy to try and add that property onto the type if there would be interest in accepting such a PR.

**Comment by jkotas** [comment]
> I'd be happy to try and add that property onto the type if there would be interest in accepting such a PR.

The first thing to do is to turn this into a format API proposal: https://github.com/dotnet/runtime/blob/master/docs/project/api-review-process.md

Also, look at the existing proposals in this space, e.g.: https://github.com/dotnet/corefx/issues/27417

**Comment by vcsjones** [comment]
> These sorts of tests normally would be done through pattern matching rather than if / thens - it's just the idiomatic way of doing things - but without a value to match on, it's a bit awkward.

Sure, but assuming you _could_ get the OSPlatform - how would the pattern match look in F#? The snippet I posted above does not compile but attempts to match on an OSPlatform. I ask because perhaps that may influence the formal API proposal.

**Comment by isaacabraham** [comment]
Understood. Let me have a bash tomorrow.

**Comment by isaacabraham** [comment]
@vcsjones as you say, as it's just a property, you can't directly pattern match against it unless you start using `when` clauses e.g.

```fsharp
let name =
    match currentPlatform with
    | _ when currentPlatform = OSPlatform.Linux -> "Linux"
    | _ when currentPlatform = OSPlatform.Windows -> "Windows"
    | _ -> "Other"
```

But assuming a single property was made available, with a simple active pattern you can then easily match on the OS:

```fsharp
let (|OS|_|) platform dest = if platform = dest then Some() else None
let name =
    match RuntimeInformation.CurrentPlatform with
    | OS OSPlatform.Linux -> "Linux"
    | OS OSPlatform.Windows -> "Windows"
    | _ -> "Other"
```

where `RuntimeInformation.CurrentPlatform` is the newly added property.

To be honest, this isn't just about F# or pattern matching if I think about it, but simply the ability to ask the application: "What operating system are you running on?".

**Comment by AnthonyMastrean** [comment]
Looks like `System.OperatingSystem` is replacing `System.Runtime.InteropServices.RuntimeInformation` as of #40111. I'm not sure where that leaves `System.Environment.OSVersion`. It doesn't look like there's any runtime constants to support pattern matching and switch expressions... and `PlatformID` is on the way out as of #55914?
